File: swipelibrary/src/main/java/kaufland/com/swipelibrary/DraggingProxy.java
package kaufland.com.swipelibrary;

import android.graphics.Rect;
import android.view.View;
import android.view.ViewGroup;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;

import kaufland.com.swipelibrary.dragengine.DraggingEngine;

import static kaufland.com.swipelibrary.SwipeLayout.LEFT_DRAG_VIEW;
import static kaufland.com.swipelibrary.SwipeLayout.RIGHT_DRAG_VIEW;
import static kaufland.com.swipelibrary.SwipeLayout.SURFACE_VIEW;
import static kaufland.com.swipelibrary.SwipeState.DragViewState.LEFT_OPEN;
import static kaufland.com.swipelibrary.SwipeViewLayouter.DragDirection.HORIZONTAL;

/**
 * Created by sbra0902 on 30.03.17.
 */
@EBean
public class DraggingProxy {

    @Bean
    protected SwipeViewLayouter mSwipeViewLayouter;

    private boolean isInitilized;


    public void init(ViewGroup parent) {
        isInitilized = true;
        mSwipeViewLayouter.init(parent);
        initInitialPosition();
    }


    public void requestLayout() {
        for (View dragView : mSwipeViewLayouter.getViews().values()) {
            dragView.forceLayout();
        }
    }

    public int clampViewPositionHorizontal(View child, int left) {

        for (DraggingEngine dragView : mSwipeViewLayouter.getViewEngines().values()) {
            if (dragView.getDragView().equals(child)) {
                return dragView.clampViewPositionHorizontal(child, left);
            }
        }
        return left;
    }

    public SwipeResult determineSwipeHorizontalState(float velocity, SwipeDirectionDetector swipeDirectionDetector, SwipeState swipeState, SwipeLayout.SwipeListener swipeListener, View child) {
        for (DraggingEngine dragView : mSwipeViewLayouter.getViewEngines().values()) {

            SwipeResult result = dragView.determineSwipeHorizontalState(velocity, swipeDirectionDetector, swipeState, swipeListener, child);
            if (result != null) {
                return result;
            }

        }
        return new SwipeResult(0);
    }

    public void initInitialPosition() {

        for (DraggingEngine dragView : mSwipeViewLayouter.getViewEngines().values()) {
            dragView.initializePosition(mSwipeViewLayouter.getDragDirection());
        }
    }

    public void moveView(View changedView, int positionChanges) {

        for (DraggingEngine engine : mSwipeViewLayouter.getViewEngines().values()) {
            engine.moveView(positionChanges, (SurfaceView) mSwipeViewLayouter.getViews().get(SURFACE_VIEW), changedView);
        }

    }

    public void restoreState(SwipeState.DragViewState state) {

        for (DraggingEngine view : mSwipeViewLayouter.getViewEngines().values()) {
            view.restoreState(state, (SurfaceView)  mSwipeViewLayouter.getViews().get(SURFACE_VIEW));
            view.getDragView().forceLayout();
        }
    }

    public SwipeViewLayouter.DragDirection getDragDirection() {
        return mSwipeViewLayouter.getDragDirection();
    }

    public boolean isCapturedViewDraggable(View child) {

        boolean draggable = false;

        for (View view : mSwipeViewLayouter.getViews().values()) {

            draggable = draggable || child.getId() == view.getId();
        }

        return draggable;
    }

    public boolean canSwipe(SwipeDirectionDetector swipeDirectionDetector, SwipeState.DragViewState state) {

        boolean canSwipe = false;

        float absDiffX = Math.abs(swipeDirectionDetector.getDifX());
        float absDiffY = Math.abs(swipeDirectionDetector.getDifY());
        float diffX = swipeDirectionDetector.getDifX();
        float diffY = swipeDirectionDetector.getDifY();
        boolean isLeftDraggable = mSwipeViewLayouter.getDragViewEngineByPosition(LEFT_DRAG_VIEW) != null ? ((DragView)mSwipeViewLayouter.getViews().get(LEFT_DRAG_VIEW)).isDraggable() : false;
        boolean isRightDraggable = mSwipeViewLayouter.getDragViewEngineByPosition(RIGHT_DRAG_VIEW) != null ? ((DragView)mSwipeViewLayouter.getViews().get(RIGHT_DRAG_VIEW)).isDraggable() : false;

        if (mSwipeViewLayouter.getDragDirection() == HORIZONTAL) {
            if (diffX > 0) {
                canSwipe = absDiffX > absDiffY && (isLeftDraggable || mSwipeViewLayouter.getSurfaceView().getX() < 0);
            } else {
                canSwipe = absDiffX > absDiffY && (isRightDraggable || mSwipeViewLayouter.getSurfaceView().getX() > 0);
            }

        }

        return canSwipe;
    }

    public boolean isInitilized() {
        return isInitilized;
    }

    public View getSurfaceView() {
        return mSwipeViewLayouter.getSurfaceView();
    }

    public int getSurfaceOpenOffsetByDragView(int dragView) {
       return mSwipeViewLayouter.getViewEngines().get(dragView).getOpenOffset();
    }

}

File: swipelibrary/src/main/java/kaufland/com/swipelibrary/DraggingProxy.kt
package kaufland.com.swipelibrary

import android.view.View
import android.view.ViewGroup
import kaufland.com.swipelibrary.SwipeLayout.Companion.LEFT_DRAG_VIEW
import kaufland.com.swipelibrary.SwipeLayout.Companion.RIGHT_DRAG_VIEW
import kaufland.com.swipelibrary.SwipeLayout.Companion.SURFACE_VIEW


import kaufland.com.swipelibrary.SwipeViewLayouter.DragDirection.HORIZONTAL
import kotlin.math.abs

/**
 * Created by sbra0902 on 30.03.17.
 */
class DraggingProxy {

    private var mSwipeViewLayouter: SwipeViewLayouter? = null

    private var mLayoutCache: LayoutCache? = null

    var isInitialized: Boolean = false
        private set

    val dragDirection: SwipeViewLayouter.DragDirection
        get() = mSwipeViewLayouter!!.dragDirection

    val surfaceView: View
        get() = mSwipeViewLayouter!!.surfaceView


    fun init(parent: ViewGroup) {
        isInitialized = true
        mSwipeViewLayouter!!.init(parent)
        initInitialPosition()
    }


    fun requestLayout() {
        for (dragView in mSwipeViewLayouter!!.views.values) {
            dragView.forceLayout()
        }
    }

    fun clampViewPositionHorizontal(child: View, left: Int): Int {

        for (dragView in mSwipeViewLayouter!!.viewEngines.values) {
            if (dragView.dragView == child) {
                return dragView.clampViewPositionHorizontal(child, left)
            }
        }
        return left
    }

    fun determineSwipeHorizontalState(velocity: Float, swipeDirectionDetector: SwipeDirectionDetector, swipeState: SwipeState, swipeListener: SwipeLayout.SwipeListener, child: View): SwipeResult {
        for (dragView in mSwipeViewLayouter!!.viewEngines.values) {

            val result = dragView.determineSwipeHorizontalState(velocity, swipeDirectionDetector, swipeState, swipeListener, child)
            if (result != null) {
                return result
            }

        }
        return SwipeResult(0)
    }

    private fun initInitialPosition() {

        for (dragView in mSwipeViewLayouter!!.viewEngines.values) {
            dragView.initializePosition(mSwipeViewLayouter!!.dragDirection)
        }
    }

    fun moveView(changedView: View, positionChanges: Int) {

        for (engine in mSwipeViewLayouter!!.viewEngines.values) {
            engine.moveView(positionChanges.toFloat(), mSwipeViewLayouter!!.views[SURFACE_VIEW] as SurfaceView, changedView)
        }

    }

    fun restoreState(state: SwipeState.DragViewState) {

        for (view in mSwipeViewLayouter!!.viewEngines.values) {
            view.restoreState(state, mSwipeViewLayouter!!.views[SURFACE_VIEW] as SurfaceView)
            view.dragView?.forceLayout()
        }
    }

    fun isCapturedViewDraggable(child: View): Boolean {

        var draggable = false

        for (view in mSwipeViewLayouter!!.views.values) {

            draggable = draggable || child.id == view.id
        }

        return draggable
    }

    fun canSwipe(swipeDirectionDetector: SwipeDirectionDetector): Boolean {

        var canSwipe = false

        val absDiffX = abs(swipeDirectionDetector.difX).toFloat()
        val absDiffY = abs(swipeDirectionDetector.difY).toFloat()
        val diffX = swipeDirectionDetector.difX.toFloat()
        val isLeftDraggable = (mSwipeViewLayouter!!.views[LEFT_DRAG_VIEW] as DragView).isDraggable
        val isRightDraggable = (mSwipeViewLayouter!!.views[RIGHT_DRAG_VIEW] as DragView).isDraggable

        if (mSwipeViewLayouter!!.dragDirection == HORIZONTAL) {
            canSwipe = if (diffX > 0) {
                absDiffX > absDiffY && (isLeftDraggable || mSwipeViewLayouter!!.surfaceView.x < 0)
            } else {
                absDiffX > absDiffY && (isRightDraggable || mSwipeViewLayouter!!.surfaceView.x > 0)
            }
        }

        return canSwipe
    }

    fun getSurfaceOpenOffsetByDragView(dragView: Int): Int {
        return mSwipeViewLayouter!!.viewEngines.getValue(dragView).openOffset
    }

    fun captureChildrenBound() {
        mLayoutCache!!.captureChildrenBound(mSwipeViewLayouter!!.views.values)
    }

    fun restoreChildrenBound() {
        mLayoutCache!!.restoreOnLayout()
    }
}
